web/api/formatters/rrd2json.c

```diff
@@ -114,10 +138,45 @@ int rrdset2anything_api_v1(
         , long group_time
         , uint32_t options
         , time_t *latest_timestamp
+        , char *context
 ) {
     st->last_accessed_time = now_realtime_sec();
 
-    RRDR *r = rrd2rrdr(st, points, after, before, group_method, group_time, options, dimensions?buffer_tostring(dimensions):NULL);
+    RRDDIM *temp_rd = NULL;
+
+    if (context) {
+        rrdhost_rdlock(st->rrdhost);
+        RRDSET *st1;
+        rrdset_foreach_read(st1, st->rrdhost) {
+            if (strcmp(st1->context, context) == 0) {
+                // Loop the dimensions of the chart
+                RRDDIM  *rd1;
+                rrdset_rdlock(st1);
+                rrddim_foreach_read(rd1, st1) {
+                    RRDDIM *rd = mallocz(rd1->memsize);
+                    memcpy(rd, rd1, rd1->memsize);
+                    rd->id = strdupz(rd1->id);
+                    rd->name = strdupz(rd1->name);
+                    rd->state = mallocz(sizeof(*rd->state));
+                    memcpy(rd->state, rd1->state, sizeof(*rd->state));
+                    memcpy(&rd->state->collect_ops, &rd1->state->collect_ops, sizeof(struct rrddim_collect_ops));
+                    memcpy(&rd->state->query_ops, &rd1->state->query_ops, sizeof(struct rrddim_query_ops));
+#ifdef ENABLE_DBENGINE
+                    if (rd->rrd_memory_mode == RRD_MEMORY_MODE_DBENGINE) {
+                        rd->state->metric_uuid = mallocz(sizeof(uuid_t));
+                        uuid_copy(*rd->state->metric_uuid, *rd1->state->metric_uuid);
+                    }
+#endif
+                    rd->next = temp_rd;
+                    temp_rd = rd;
+                }
+                rrdset_unlock(st1);
+            }
+        }
+        rrdhost_unlock(st->rrdhost);
+    }
```

Comment:
That's yet another huge race condition we're adding to the agent. We're cloning dimension structures. The original dimension structures could be freed after that, or the internal structures they point to.

Reply:
> That's yet another huge race condition we're adding to the agent. We're cloning dimension structures. The original dimension structures could be freed after that, or the internal structures they point to.

True and if not an immediate danger from the structures since they are cloned, they need the rrdset (or a clone of) :smile: